pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,223 @@
+            metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: MetricsAgent) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
+            if (time.perf_counter() - pulled_at) >= ack_deadline:
+                metrics_client.increment('pubsub.consumer.failfast')
+                message_queue.task_done()
+                semaphore.release()
+                continue
+
+            metrics_client.histogram(
+                'pubsub.consumer.latency.receive',
+                # publish_time is in UTC Zulu
+                # https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
+                time.time() - message.publish_time.timestamp())
+
+            asyncio.ensure_future(_fire_callback(
+                semaphore,
+                message,
+                callback,
+                ack_queue,
+                metrics_client,
+            ))
+            message_queue.task_done()
+
+    async def producer(
+            subscription: str,
+            message_queue: MessageQueue,
+            subscriber_client: 'SubscriberClient',
+            max_messages: int,
+            metrics_client: MetricsAgent
+    ) -> None:
+        while True:
+            try:
+                new_messages = await subscriber_client.pull(
+                    subscription=subscription, max_messages=max_messages)
+            except asyncio.TimeoutError:
+                continue
+
+            pulled_at = time.perf_counter()
+            for m in new_messages:
+                await message_queue.put((m, pulled_at))
+
+            metrics_client.histogram(
+                'pubsub.producer.batch', len(new_messages))
+
+            await message_queue.join()
+
+    async def subscribe(subscription: str,  # pylint: disable=too-many-locals
+                        handler: ApplicationHandler,
+                        *,
+                        num_workers: int = 1,
+                        max_messages: int = 100,
+                        ack_window: float = 0.3,
+                        ack_deadline_cache_timeout: int = 10,
+                        consumer_pool_size: int = 1,
+                        subscriber_client: Optional[SubscriberClient] = None,
+                        metrics_client: Optional[MetricsAgent] = None
+                        ) -> None:
+        ack_queue: 'asyncio.Queue[str]' = asyncio.Queue(
+            maxsize=(max_messages * num_workers))
+        subscriber = subscriber_client or SubscriberClient()
```

Comment:
Ah, this would be the cause of that unclosed session issue you might be running into -- in cases where users don't provide a `SubscriberClient`, the one you create here never explicitly calls `.close()` to properly shutdown its sessions (see the whole `__aenter__` / `__aexit__` thing).

Honestly, maybe it'd be easiest to just make `subscriber_client` a *non-optional parameter* to enforce proper usage. Thoughts?

Reply:
I actually... like this very very much. Like, don't be lazy, spend one line and pass me that client, we're doing _composition_ here.